### PR TITLE
Feature/hashed passwords

### PR DESCRIPTION
**Pull request by louismunro**
## Description

Allows hashing the passwords in temporary_password with bcrypt.
Support for additional hashing algorithm should be easy as the code is already meant for that although a bit of refactoring wouldn't hurt.
## Impacts

Have a look at pf::password, it renames  pf::temporary_password.
Test authentication with a database user.
Test changing the password as well as the algorithm. 
Test backwards compatibility with plaintext. 

Test captive portal authentication.
Test the configurator.
Test option to send password reminders to users.
Test option to print passwords.
## NEWS file entries

New Features
++++++++++++

Enhancements
++++++++++++
- Database passwords can now be hashed with bcrypt.

Bug Fixes
+++++++++
## Delete branch after merge

YES

### CODE REVIEW

File: html/pfappserver/lib/pfappserver/PacketFence/Controller/Configurator.pm [hunk 1]
```diff
@@ -292,37 +292,45 @@ sub configuration :Chained('object') :PathPart('configuration') :Args(0) {
...
-        unless ($general_domain && $general_hostname && $general_dhcpservers && $alerting_emailaddr) {
+        my $general_domain          = $c->request->params->{'general.domain'};
+        my $general_hostname        = $c->request->params->{'general.hostname'};
+        my $general_dhcpservers     = $c->request->params->{'general.dhcpservers'};
+        my $alerting_emailaddr      = $c->request->params->{'alerting.emailaddr'};
+        my $advanced_hash_passwords = $c->request->params->{'advanced.hash_passwords'};
+
+        unless ($general_domain && $general_hostname && $general_dhcpservers &&
+                $alerting_emailaddr &&
+                $advanced_hash_passwords) {
```
Comment (julsemaan) [inline html/pfappserver/lib/pfappserver/PacketFence/Controller/Configurator.pm]: $alerting_emailaddr and $advanced_hash_passwords could go on the same line here

Comment (louismunro) [inline html/pfappserver/lib/pfappserver/PacketFence/Controller/Configurator.pm]: Actually the whole condition is pretty ugly.
Let me see if I can make it better.

Comment (cgx) [inline html/pfappserver/lib/pfappserver/PacketFence/Controller/Configurator.pm]: The idea was to group the parameters by section (_general_, _alerting_, and _advanced_).

File: html/pfappserver/root/configurator/configuration.tt [hunk 2]
```diff
@@ -53,6 +53,20 @@
                 <p class="help-block">[% l('Email address to which notifications for rogue DHCP servers, violations with an action of <i>email</i>, or any other PacketFence-related messages goes to.') %]</p>
               </div>
             </div>
+            <h3>[% l('Database Passwords') %]</h3>
```
Comment (julsemaan) [inline html/pfappserver/root/configurator/configuration.tt]: I see database password as ambiguous.
What do you think of 'Local passwords'

File: html/pfappserver/root/static/configurator/configuration.js [hunk 3]
```diff
@@ -32,7 +32,8 @@ function saveStep(href) {
             data: {'general.domain': $('#general_domain').val(),
                    'general.hostname': $('#general_hostname').val(),
                    'general.dhcpservers': $('#general_dhcpservers').val(),
-                   'alerting.emailaddr': $('#alerting_emailaddr').val()}
+                   'alerting.emailaddr': $('#alerting_emailaddr').val(),
+                   'advanced.hash_passwords': $('input[name="advanced.hash_passwords"]').val()}
```
Comment (julsemaan) [inline html/pfappserver/root/static/configurator/configuration.js:36]: All the other ones are using an html ID. Would be worth it to make it the same for all the fields in this section

File: html/captive-portal/lib/captiveportal/PacketFence/Controller/Signup.pm [hunk 4]
```diff
@@ -5,7 +5,7 @@ use Date::Format qw(time2str);
 use pf::log;
 use pf::constants;
 use pf::config;
-use pf::temporary_password 1.11;
+use pf::password 2.01;
```
Comment (rouzier) [inline html/captive-portal/lib/captiveportal/PacketFence/Controller/Signup.pm]: I would just drop the version number to does not really give us anything and it is one more thing to maintain

File: html/pfappserver/root/configurator/configuration.tt (hunk 2 above)
Comment (louismunro) [inline html/pfappserver/root/configurator/configuration.tt]: Not sure it's any less ambiguous.
To an uneducated user (which is most people when they use the configurator) it is not much help.
Local database passwords?

File: html/pfappserver/root/static/configurator/configuration.js (hunk 3 above)
Comment (cgx) [inline html/pfappserver/root/static/configurator/configuration.js:36]: This is different, it's a radio button. Each input could have a different ID. We must rely on the name attribute.

File: html/captive-portal/lib/captiveportal/PacketFence/Controller/Signup.pm (hunk 4 above)
Comment (louismunro) [inline html/captive-portal/lib/captiveportal/PacketFence/Controller/Signup.pm]: Agreed. Consider it done.

File: lib/pf/password.pm [hunk 5]
```diff
@@ -447,9 +443,98 @@ sub validate_password {
...
+    # Plaintext passwords have no prefix.
+    # We need to quotemeta the regex because it contains { and }
+    my $bcrypt_re = quotemeta('{bcrypt}');
+    my $md5_re    = quotemeta('{md5}');
+    my $nthash_re = quotemeta('{NT}');
+
+    switch ($hash_string) {
+        case /$bcrypt_re/ { return _check_bcrypt(@_) }
+        case /$md5_re/    { return undef }
+        case /$nthash_re/ { return undef }
```
Comment (rouzier) [inline lib/pf/password.pm]: EIther implement nthash and md5 or disable them in _hash_password.

File: lib/pf/web/guest.pm [hunk 6]
```diff
@@ -46,7 +46,7 @@ BEGIN {
 
 use pf::constants;
 use pf::config;
-use pf::temporary_password 1.11;
+use pf::password 2.01;
```
Comment (rouzier) [inline lib/pf/web/guest.pm]: Remove version number

File: lib/pf/password.pm [hunk 7]
```diff
@@ -447,9 +443,98 @@ sub validate_password {
...
+    # the algorithm is contained in the prefix of the password such as
+    # {md5}, {bcrypt} etc.
+    # Plaintext passwords have no prefix.
+    # We need to quotemeta the regex because it contains { and }
+    my $bcrypt_re = quotemeta('{bcrypt}');
+    my $md5_re    = quotemeta('{md5}');
+    my $nthash_re = quotemeta('{NT}');
+
+    switch ($hash_string) {
+        case /$bcrypt_re/ { return _check_bcrypt(@_) }
```
Comment (julsemaan) [inline lib/pf/password.pm]: if it's a prefix, we should probably use /^$brypt_re/ to make it more explicit

File: lib/pf/password.pm [hunk 8]
```diff
@@ -447,9 +443,98 @@ sub validate_password {
...
+        case /$md5_re/    { return undef }
```
Comment (julsemaan) [inline lib/pf/password.pm]: Same here

File: lib/pf/password.pm (hunk 5 above)
Comment (julsemaan) [inline lib/pf/password.pm]: Same here

File: lib/pf/password.pm [hunk 9]
```diff
@@ -447,9 +443,98 @@ sub validate_password {
...
+sub _hash_password {
+    my ( $plaintext, %params ) = @_;
+
+    switch ( $params{"algorithm"} ) {
+        case 'plaintext' { return $plaintext }
+        case 'bcrypt'    { return bcrypt( $plaintext, %params ) }
+        case 'nthash'    { return nthash $plaintext }
+        case 'md5'       { return md5 $plaintext }
+        else {
+            logger->error( "Unsupported hash algorithm " . $params{"algorithm"} );
```
Comment (julsemaan) [inline lib/pf/password.pm:474]: logger ? get_logger maybe ?

File: lib/pf/password.pm [hunk 10]
```diff
@@ -447,9 +443,98 @@ sub validate_password {
...
+        }
+    }
+}
+
+sub nthash { return undef; }    # TODO
```
Comment (julsemaan) [inline lib/pf/password.pm]: A little logging message with that ?

File: lib/pf/password.pm [hunk 11]
```diff
@@ -447,9 +443,98 @@ sub validate_password {
...
+
+sub md5 { return undef; }       # TODO
```
Comment (julsemaan) [inline lib/pf/password.pm]: logging here too ?

File: lib/pf/password.pm [hunk 12]
```diff
@@ -460,12 +545,17 @@ sub reset_password {
     # Making sure pid/password are "ok"
     if ( !defined($pid) || !defined($password) || (length($pid) == 0) || (length($password) == 0) ) {
         $logger->error("Error while resetting the user password. Missing values.");
-        return;
+        return undef;
     }
 
+    # hash the password if required
+    if ( $Config{'advanced'}{'hash_passwords'} ne 'plaintext' ) {
```
Comment (julsemaan) [inline lib/pf/password.pm]: plaintext should be moved into a constant

File: lib/pf/password.pm [hunk 13]
```diff
@@ -447,9 +443,98 @@ sub validate_password {
...
+    my $prefix_len        = 12;
+    my $cost_len          = 2;
+    my $salt_len          = 22;
+    my $before_salt       = $prefix_len + $cost_len + 1;    # +1 for the trailing $
+    my $before_hash_value = $before_salt + $salt_len;
+    $cost = substr( $hash_string, $prefix_len,  $cost_len );
+    $salt = substr( $hash_string, $before_salt, $salt_len );
+    $hash_value = substr( $hash_string, $before_hash_value );    # substr to the end of the string
+
+    $hashed_plaintext = _hash_password( $plaintext, algorithm => 'bcrypt', salt => $salt, cost => $cost );
```
Comment (julsemaan) [inline lib/pf/password.pm]: brypt should be moved into a constant

File: lib/pf/password.pm [hunk 14]
```diff
@@ -447,9 +443,98 @@ sub validate_password {
...
+    }
+    else {
+        return $FALSE;
+    }
+}
+
+sub bcrypt {
+    my ( $plaintext, %params ) = @_;
+
+    use Data::Entropy::Algorithms qw( rand_bits );
```
Comment (julsemaan) [inline lib/pf/password.pm]: This use should be moved at the top of the file

File: lib/pf/password.pm [hunk 15]
```diff
@@ -447,9 +443,98 @@ sub validate_password {
...
+    use Crypt::Eksblowfish::Bcrypt qw(bcrypt_hash en_base64 de_base64 );
```
Comment (julsemaan) [inline lib/pf/password.pm]: This one too

File: lib/pf/password.pm (hunk 14 above)
Comment (julsemaan) [inline lib/pf/password.pm]: Also add this to the builds spec files

File: lib/pf/password.pm (hunk 15 above)
Comment (julsemaan) [inline lib/pf/password.pm]: Needs to be added to the builds

File: html/pfappserver/root/configurator/configuration.tt (hunk 2 above)
Comment (julsemaan) [inline html/pfappserver/root/configurator/configuration.tt]: that seems better, let's go for Local database passwords

File: lib/pf/password.pm (hunk 12 above)
Comment (julsemaan) [inline lib/pf/password.pm]: You forgot me

**Comment by louismunro** [comment]
Merge me if ready please.

**Comment by julsemaan** [comment]
Only thing is this needs everyone to upgrade their schema. I think we should be asking someone to briefly test this from scratch (devel -> this) just so we're not all locked out of our admins and getting all sort of auth errors on our setups.

**Comment by louismunro** [comment]
I have done that already.
I will do it again with the latest devel.  

## 

LM

**Comment by rouzier** [comment]
Have a htpassword source as a backup

**Comment by louismunro** [comment]
You can always edit the database directly.
Plaintext passwords are still accepted for backwards compatibility.

## 

LM

**Comment by julsemaan** [comment]
Works for me then, I'll merge

**Pull request merged**